src/data/nat/factorial.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2018 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Chris Hughes, Floris van Doorn
+-/
+import data.nat.basic
+/-!
+# The factorial function
+
+-/
+
+namespace nat
+variables {m n : ℕ}
+
+/-- `nat.factorial n` is the factorial of `n`. -/
+@[simp] def factorial : nat → nat
+| 0        := 1
+| (succ n) := succ n * factorial n
+
+localized "notation n `!`:10000 := nat.factorial n" in nat
```

Comment:
```suggestion
localized "local notation n `!`:10000 := nat.factorial n" in nat
```
I think that otherwise it gets injected as global notation as soon as someone opens the locale. But maybe `localized` takes care of this? @fpvandoorn

Reply:
I think `localized` implies `local`, since trying to do what you suggested yields an error. Also, a quick look at the rest of mathlib seems to show that all localized notations are declared like this.